### ISSUE TITLE
how to handle credits for translation?

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/joist#163]
AboutDialog currently reads credits.translation for translation credits. But credits is currently specified in a sim's main.js, and so that's not an appropriate place to specify translation credits.  Translation credits should be specified in the string files. But a sim will have multiple string files (sim-specific, common-code,...) so we need to work out how to coalesce those into a translation credit.

**Comment by pixelzoom** [comment]
At 2/17/15 UX design meeting with @ariel-phet, the proposal was to credit translators on the PhET website, but not in the About dialog. This allows us to avoid the issue of how to credit multiple translators, sim vs common-code translators, etc.  Assigning to @ariel-phet to clear this with @kathy-phet.

**Comment by pixelzoom** [comment]
@ariel-phet bump

**Comment by ariel-phet** [comment]
I will see if I can catch @kathy-phet 's ear this week.

**Comment by kathy-phet** [comment]
I am feeling a bit uneasy with removing the credit from the sim, since these translators do do a lot for us for free, and giving visible credit is a nice acknowledgement of that important contribution. We saw that in our groups own response to changing the credit dialog. 

I would consider a compromise (describe below), depending on feedback from the translators. 

The propose compromise would be to provide some text and an easy way to link from the sim back to the website to view the translator credits. This link can be of a generic or standard format, and the updates/changes/tracking of credits happens on the website end. 

We could consider going one step further (depending on the complexity), and the sim could potentially pull information from some standard place on the website (if online) or not (if not online).

I would also like to ping our most consistent translators to get their input on the options and how important credit in the sim is to them. I want to make sure they would be OK with this approach. If named credit directly in the sim is really important to them, we will need to reconsider.

**Comment by ariel-phet** [comment]
@kathy-phet and I discussed this briefly, and although not clear from our original description the plan was to always have a link in the sim to translation credits. We obviously value our translators and want to credit them, it just might cause a fair bit of complexity to keep that credit up to date within the sim itself. More discussion to come...

**Comment by pixelzoom** [comment]
Note: Provide ability to be "anonymous" wrt translation credits. Perhaps specify this when they create their login.

**Comment by pixelzoom** [comment]
Notes from 3/5/15 status meeting discussion:

- @oliver-phet will ping key translators to see how they feel about this issue.
- No consensus on what the website "translation credits" should look like, but...
- One idea for website: list all translators by locale, in alphabetical order, click on their name to see details about them and what specific things they contributed to (sim names & "common code").
- Provide ability to be "anonymous" wrt translation credits when they create their login.

**Comment by pixelzoom** [comment]
Since this is related to translation, it should be resolved (or at least reviewed) for "Summer 2015 redeploy".  Note that it is still assigned to @ariel-phet.

**Comment by ariel-phet** [comment]
@oliver-phet - have we sent an email to a few translators yet to see if they would be OK with a webpage linked from the sim crediting them?

**Comment by oliver-phet** [comment]
I sent an email to our trusted translators 2 months ago about credits. We talked about it in the translation meeting, but basically, yes they are fine with a webpage linked from the sim about dialog.

**Comment by oliver-phet** [comment]
![unnamed](https://cloud.githubusercontent.com/assets/5090275/7991202/a71c20d6-0ab3-11e5-911f-ba0383986aff.png)


Clicking the "Translator Credits" will take the user to the translation credits page that will have all of the translators listed by locale, in alphabetical order. The translator's name can be clicked to see details about them (including website URLs) and what specific portions of the PhET project they contributed translations to (sim name(s) and "common code").

Because the new HTML5 utility will require the translator to log into the PhET website to submit a translation, the credits webpage will be populated with new information we collect from a translator's user name. The translator will be able to decide what credit information they would like to be shown and edit that information at any time. We will also provide the ability for a translator to be "anonymous".

**Comment by oliver-phet** [comment]
Assigning this back to @ariel-phet

**Comment by ariel-phet** [comment]
This issue is basically going to be handled by https://github.com/phetsims/website/issues/224

**Comment by pixelzoom** [comment]
The "Translation Credits" link in the About dialog needs to be done.

**Comment by pixelzoom** [comment]
7/9/15 dev meeting, do these tasks ASAP:

• Add "Translation Credits" link to the About dialog. (Note the name difference, not "Translator Credits" as in the mockup above.)
• Put up the associated webpage page.

**Comment by zepumph** [comment]
There are now two links on the about dialog, one for translation credits, and one for
licensing. Both link to the dev website, where the translation credit page is live. This will need to be changed manually when it is deployed.

**Comment by pixelzoom** [comment]
@zepumph I don't see your commits referenced in this issue.  It's standard practice to include the issue number in your commit messages, see https://help.github.com/articles/writing-on-github/#references.

You said:
> Both link to the dev website, where the translation credit page is live. This will need to be changed manually when it is deployed.

Why does the license link go to the dev website? It was going to a valid URL on the PhET website that was going to be supported by a redirect.

**Comment by samreid** [comment]
> There are now two links on the about dialog...

I actually saw 3 links on the about dialog, please see #274

**Comment by pixelzoom** [comment]
[I moved these comments here from #274.]

@zepumph A few of things related to this...

(1) I suspect that you're not running with the 'phet' branch of the brand repository, which is why you didn't see the existing license link. If you run with the 'master' branch of brand, you'll be using the default (empty) brand, which currently has no links field in Brand.js.

(2) Speaking of the "links" field in Brand.js... Notice that the license link is handled in Brand.js. I suspect that the Translation Credits link may also need to be brand-specific too. If that's the case, then inserting the sim name into the URL may be a little trickier than what you've done here.

(3) af931a6 shows reformatting of code in AboutDialog.js that is not related to this issue (around line 69, children.push(...). Are you using the PhET code style in IntelliJ IDEA?

(4) AboutDialog.js line 121 needs to be removed:

console.log(devName);

The lack of spacing around parens is another clue that you're probably not using the PhET code style.

(5) AboutDialog.js line 105 has a private static function declared inside the constructor:

var displayNameToDevName = function(...

This will cause a copy of this function to be created for every instance of this type. In general, declare private static functions outside of the constructor to conserve memory.

**Comment by pixelzoom** [comment]
Recommendations:

1. Revert all changes to joist

2. Move the Translation Credits link to Brand.js for the brands that should have such credits.

3. Remove the '#simname' portion of the Translation Credits URL, so that this link can be handled through the normal "brand links" mechanism in AboutDialog.  Imo, scrolling that URL to a sim-specific place would be nice, but is not essential, and introduces special-case handling of links that should be avoided.  So the cost is not worth the benefit. 

@zepumph and @aaronsamuel137, it would be a good idea to confirm the above with @ariel-phet and seek input from other developers before proceeding.

**Comment by pixelzoom** [comment]
Looks like these are the joist commits related to this issue: 
https://github.com/phetsims/joist/commit/af931a6ebf8f831c1857d485421b0c1240286df3
https://github.com/phetsims/joist/commit/6588ffbe1eb514576c6ef6e414ddcd8bcfaeda47
https://github.com/phetsims/joist/commit/c032e9cc9fbb1b56cea380e07f480e5852d4869f

Where any other repos involved?

**Comment by zepumph** [comment]
Hey there, sorry about not referencing the commits, I usually do. Those were the only three, no other repos involved. 

I linked the license to phet-dev because hopefully soon (still waiting on a mock up) we are going to get the new licensing pages up on dev, so then we would see the better version of the licensing structure.

As for the list of items for me, thank you for taking to time to write out the advice, I really appreciate the help! 

1) I had no idea about the inner workings of phet/brand, being new to the sim side and all, but it makes sense and I'll checkout the 'phet' branch before continuing. 

2) I will remove the licensing link I added, sorry for the confusion, I just went off of the mock up I saw in an earlier comment, and didn't really think about how it didn't pertain to this issue. I will keep commits more specific from here on. I'm not sure if translation credits will need to be brand specific, but if it is determined so, then I'll move the link into brand and we will make the link simply point to the top of the page, no #.

3/4) I am using the phet code style, I usually reformat the code before committing. In this case, I added the console.log for debugging, and then forgot to remove it, and also forgot to reformat again before committing. I will work on cleaning up code before commits. 

5) That is something that is obvious to me in retrospect, just a stupid mistake i guess. Sorry about that. I'll move it out of the constructor, if we decide to keep the translation link in the about dialog. 

I'm sorry for all of the mistakes, and the confusion, I'll work on it. Thanks for your patience.

**Comment by pixelzoom** [comment]
No problem. And as I said, make sure you get @ariel-phet and other developers to weigh in before implementing my recommendations.

**Comment by zepumph** [comment]
@ariel-phet  we are looking for your thoughts on where to put the translation credit link in the About Dialog of each sim. Chris is thinking it may be good to put it in brand, because we may want it to disappear for certain brands like the adapted from phet. This is along the same lines as issue phetsims/brand#13. 

The upside to keeping it in the About dialog is that it is much easier to implement the url taking you to #the-sim-name-you-are-looking-for instead of just the top of the page, like we would more likely do if it gets put into Brand.js. 

Assigning to @ariel-phet for thoughts, feelings, and the guidance he is so good at giving.

**Comment by aaronsamuel137** [comment]
@ariel-phet - Also please comment on whether or not you think it is acceptable to remove the ```#simname-header``` portion of the url. This is @pixelzoom's recommendation to simplify the code if we do end up putting this link in brand, but it would mean getting rid of the feature where you automatically get scrolled down to the sim you are linking from.

We can also discuss at developer meeting if you think that would be more useful.

**Comment by zepumph** [comment]
When I talked to @pixelzoom this morning he made it seem like there were multiple items that should be talked about in correlation with which brand we should put them in. So it may be good to get more input on this topic at dev-meeting.

**Comment by ariel-phet** [comment]
dev-meeting it is...I am leaning towards @pixelzoom suggestion.

I agree that scrolling to the sim is not essential. We want to make sure we are giving translators credit, but someone motivated will easily be able to go to the appropriate sim from the link, even if not autoscrolled, so I agree this functionality though nice, is not essential. And if avoiding the functionality makes for cleaner code and less potential maintenance headaches, I think that makes for a reasonable argument against.

 In addition, it is true that likely we will want to remove translation credits from certain brands.

**Comment by pixelzoom** [comment]
Looks like the 2nd license link was removed in the commits:

https://github.com/phetsims/joist/commit/e2d2a5bf6b61a58e5854cfe489876baac59d93f5
https://github.com/phetsims/joist/commit/d5fabbb670f833bfca351d6c4b01e65b24c7497b
https://github.com/phetsims/joist/commit/3b3b1a4cc2e026bb5aa35317556eb8a8a71845e2

Let's try to get the issue number in commit messages.

**Comment by zepumph** [comment]
As discussed at 7/16 dev meeting:

Move  the translation credit link into phet brand, sacrificing the #sim-name functionality.

**Comment by pixelzoom** [comment]
Translation Credits link also need to be in interoperable brand. I added it.

**Comment by samreid** [comment]
In my opinion, it would not be too complex to add a new set of "dynamic" brand links, which is a `function that returns an array of strings` rather than an `array of strings`.  The function would take in sim metadata as an argument (perhaps the Sim instance) and could use it to create custom links.

I'm not advocating doing this *now*, just wanted to brainstorm what it could look like if we decide we need it, and so we are on the same page about the upper limit of the complexity of this kind of feature.

**Comment by kathy-phet** [comment]
We discussed this today at Developer Meeting, and new ideas emerged as to how to structure the Translation Credit url and build to link to the Sim-specific table in a way that is still nice code and maintainable. 

I am in favor of this upgrade - as linking just to the translation credit page in general makes it much harder to find the correct translation (a number of scrolls and clicks). If someone is looking for the translator of their sim or just wants to see who / what organization is translating for them, I think making the link jump to the table with the table open is a pretty big benefit, and worth the effort (if code complexity is maintained as reasonable.)

**Comment by samreid** [comment]
At the developer meeting, @aaronsamuel137 @jbphet @jessegreenberg @zepumph and @oliver-phet and @samreid and @jonathanolson collaborated on this and decided to push the above commits.  We still need to handle the Brand branches.

**Comment by samreid** [comment]
Finished the branches above.  Reassigning to @jessegreenberg to review everything.

**Comment by aaronsamuel137** [comment]
Also the changes made to Brand will require changes to the website. We'll need to redirect ```/translation-credits?simName=area-builder&locale=fr``` to ```/fr/for-translators/translation-credit#area-builder```.

The redirect is useful for 2 reasons:
1. The locale might not exist on the website, so we want to check if it exists before going to a non-existent URL.
2. It will be easier to change the url in the future if we need to.

**Comment by samreid** [comment]
Thanks to @jessegreenberg for pointing out a casing error in the packageJSON.js filename, I have fixed it above.

**Comment by samreid** [comment]
@jessegreenberg when your review is complete, please reassign it to @pixelzoom for further review and discussion.

**Comment by jessegreenberg** [comment]
I consider my review compete.  Code looks good, and I tested all branches with molecule-shapes in both requirejs and built versions.  Reassigning to @pixelzoom for further review.

**Comment by pixelzoom** [comment]
I took it for a test drive. Clicking on the "Translation Credits" link in the About dialog, the URL that's opened in the browser looks correct. (Eg, http://phet.colorado.edu/en/for-translators/translation-credit#acid-base-solutions).  

One minor issue: The title on the web page says "Translation Credit", and it should be "Translation Credits" (plural).  I don't know if the web work was handled via a separate issue, so I'll assign this to @aaronsamuel137 for followup.

**Comment by zepumph** [comment]
I just changed the title to translation credits, It will take a bit to show up because of varnish. Were you thinking that the url should also be '/translation-credits#sim , or just the Title? @pixelzoom

**Comment by pixelzoom** [comment]
I hadn't noticed that the URL was singular, it should be plural also.

**Comment by aaronsamuel137** [comment]
Thanks for taking care of this @zepumph. Looks like the url will be updated on the next deploy. I also made the side nav link also say "Translation Credits" instead of "Translation Credit".

Anything else that needs to be done here or can we close?

**Comment by pixelzoom** [comment]
For consistency, how about renaming TranslationCreditPanel.java to TranslationCreditsPanel.java?

I'm also a little confused... I synced with the latest joist change to get the updated URL, and I'm still being taken to `http://phet.colorado.edu/en/for-translators/translation-credit#hookes-law`.  I would except to see either the new URL or "not found" until the website is next deployed.

**Comment by zepumph** [comment]
The change is from the website side, and isn't deployed to production yet. The changes are only pushed to master. They are on phet-dev though. If you go to the same URL.

**Comment by pixelzoom** [comment]
If you want me to test, let me know when the website is redeployed. Or just go ahead and close this.

**Comment by aaronsamuel137** [comment]
Deployed and tested. Closing.

**Comment by samreid** [comment]
There are still TODOs marked for this issue, discovered during https://github.com/phetsims/chipper/issues/946

**Comment by samreid** [comment]
The TODO marked for this issue is here:

```js
    //TODO obtain translation credit from strings file, see https://github.com/phetsims/joist/issues/163
    // Translation
    if ( credits.translation ) {
      if ( children.length > 0 ) { children.push( new VStrut( 13 ) ); }
      children.push( new Text( creditsTranslationString, {
        font: options.titleFont,
        tagName: 'h2',
        innerContent: creditsTranslationString
      } ) );
      children.push( new RichText( credits.translation, richTextOptions ) );
    }
```

Some questions:
1. Why is credits.translation the conditional for this?  Is it ever true?  I don't see any "translation:" hits in `*main*` files.
2. Why is credits taken from both `creditsTranslationString` and `credits.translation`?

@jbphet can you please comment on these questions, and how translation credits are currently handled?

**Comment by jbphet** [comment]
Let me briefly comment on what I know off the top of my head, and you can send it back to me if you need more information.  We don't build translation credits into the "About" dialog, we just add a link to the website page the tracks translation credits.  Those credits are built from information in the translated string files in the babel repo.

**Comment by samreid** [comment]
Based on @jbphet's remark, it sounds like we could delete the TODO and close this issue, since we have a working solution for translator credits that doesn't require them to be shown in the about dialog.  @jbphet if you agree, can you please remove the TODO and close this issue, or let me know if I should do so?

**Comment by jbphet** [comment]
> Based on @jbphet's remark, it sounds like we could delete the TODO and close this issue...

As far as I can tell, and from what I know, I don't think that this code is ever used.  I think it may be some work that was started but never completed, and that was intended to incorporate translation credits directly into the sim rather than having a link to translation credits on the website.  I think the whole `if` block should be removed and not just the TODO.

**Comment by jbphet** [comment]
@samreid - I went ahead and removed the code as described in the comment just above.  Since this is more than you had recommended, I like to get your sign-off that this seems reasonable, so I'm assigning it back to you for review.

**Comment by samreid** [comment]
I ran several sims in aqua to test whether `credits.translation` was ever defined, and it was always `undefined` in the cases I tested.  Thanks, closing!
